File: imperative/java-imperative/src/main/java/com/fernandoglizt/imperative/javaimperative/model/ProcessResponse.java
package com.fernandoglizt.imperative.javaimperative.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.Instant;

public record ProcessResponse(
    boolean ok,
    int count,
    int batch,
    @JsonProperty("io_delay_ms") int ioDelayMs,
    @JsonProperty("downstream_url") String downstreamUrl,
    @JsonProperty("processed_events") int processedEvents,
    int batches,
    @JsonProperty("external_calls_ok") int externalCallsOk,
    @JsonProperty("external_calls_fail") int externalCallsFail,
    @JsonProperty("duration_ms") long durationMs,
    @JsonProperty("events_per_sec") double eventsPerSec,
    Instant ts,
    String node,
    String notes
) {
    public static ProcessResponse success(
        ProcessRequest request,
        int processedEvents,
        int batches,
        int externalCallsOk,
        int externalCallsFail,
        long durationMs,
        String hostname
    ) {
        double eventsPerSec = durationMs > 0 ? (processedEvents * 1000.0 / durationMs) : 0.0;
        
        return new ProcessResponse(
            true,
            request.count(),
            request.batch(),
            request.getIoDelayMs(),
            request.getDownstreamUrl(),
            processedEvents,
            batches,
            externalCallsOk,
            externalCallsFail,
            durationMs,
            eventsPerSec,
            Instant.now(),
            hostname,
            "imperative blocking with Virtual Threads (Java 21)"
        );
    }
}

File: imperative/kotlin-imperative/src/main/kotlin/com/fernandoglizt/imperative/kotlinimperative/model/ProcessResponse.kt
package com.fernandoglizt.imperative.kotlinimperative.model

import com.fasterxml.jackson.annotation.JsonProperty
import java.time.Instant

data class ProcessResponse(
    val ok: Boolean,
    val count: Int,
    val batch: Int,
    @JsonProperty("io_delay_ms") val ioDelayMs: Int,
    @JsonProperty("downstream_url") val downstreamUrl: String,
    @JsonProperty("processed_events") val processedEvents: Int,
    val batches: Int,
    @JsonProperty("external_calls_ok") val externalCallsOk: Int,
    @JsonProperty("external_calls_fail") val externalCallsFail: Int,
    @JsonProperty("duration_ms") val durationMs: Long,
    @JsonProperty("events_per_sec") val eventsPerSec: Double,
    val ts: Instant,
    val node: String,
    val notes: String
) {
    companion object {
        fun success(
            request: ProcessRequest,
            processedEvents: Int,
            batches: Int,
            externalCallsOk: Int,
            externalCallsFail: Int,
            durationMs: Long,
            hostname: String
        ): ProcessResponse {
            val eventsPerSec = if (durationMs > 0) processedEvents * 1000.0 / durationMs else 0.0
            
            return ProcessResponse(
                ok = true,
                count = request.count,
                batch = request.batch,
                ioDelayMs = request.getIoDelayMsNormalized(),
                downstreamUrl = request.getDownstreamUrlOrDefault(),
                processedEvents = processedEvents,
                batches = batches,
                externalCallsOk = externalCallsOk,
                externalCallsFail = externalCallsFail,
                durationMs = durationMs,
                eventsPerSec = eventsPerSec,
                ts = Instant.now(),
                node = hostname,
                notes = "imperative blocking with Kotlin coroutines async/await"
            )
        }
    }
}

